<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>proto_err.align &mdash; proto_err 0.0.1 documentation</title>
    
    <link rel="stylesheet" href="../../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '0.0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <link rel="top" title="proto_err 0.0.1 documentation" href="../../index.html" />
    <link rel="up" title="Module code" href="../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../../index.html">proto_err 0.0.1 documentation</a> &raquo;</li>
          <li><a href="../index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <h1>Source code for proto_err.align</h1><div class="highlight"><pre>
<span class="c">#!/usr/bin/env python</span>
<span class="c"># Aligning reads to a reference</span>
<span class="c"># Mostly wrappers for samtools scripts</span>

<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">Bio.Sequencing</span> <span class="kn">import</span> <span class="n">Applications</span>
<span class="kn">from</span> <span class="nn">Bio.Application</span> <span class="kn">import</span> <span class="n">_Option</span><span class="p">,</span> <span class="n">_Argument</span><span class="p">,</span> <span class="n">_Switch</span><span class="p">,</span> <span class="n">AbstractCommandline</span><span class="p">,</span><span class="n">_StaticArgument</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="k">def</span> <span class="nf">refIndex</span><span class="p">(</span><span class="nb">file</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to generate BWA index</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Creating BW index of reference&quot;</span><span class="p">)</span>
    <span class="n">index_cmd</span> <span class="o">=</span> <span class="n">Applications</span><span class="o">.</span><span class="n">BwaIndexCommandline</span><span class="p">(</span><span class="n">infile</span><span class="o">=</span><span class="nb">file</span><span class="p">,</span> <span class="n">algorithm</span><span class="o">=</span><span class="s">&quot;bwtsw&quot;</span><span class="p">)</span>
    <span class="n">index_cmd</span><span class="p">()</span>
    <span class="k">return</span> <span class="mi">1</span>

<span class="k">def</span> <span class="nf">align</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="n">read_file</span><span class="p">,</span> <span class="n">stdout</span><span class="p">,</span><span class="n">algorithm</span><span class="o">=</span><span class="s">&#39;bwa-mem&#39;</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">algorithm</span><span class="o">==</span><span class="s">&#39;bwa-mem&#39;</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&quot;Aligning reads to reference with bwa-mem&quot;</span><span class="p">)</span>
        <span class="n">alignCmd</span> <span class="o">=</span> <span class="n">BwaMemAlignCommandline</span><span class="p">(</span> <span class="n">reference</span><span class="o">=</span><span class="n">reference</span><span class="p">,</span> <span class="n">read_file</span><span class="o">=</span><span class="n">read_file</span><span class="p">)</span>
        <span class="k">print</span> <span class="n">alignCmd</span>

    <span class="k">return</span> <span class="n">alignCmd</span><span class="p">(</span><span class="n">stdout</span><span class="o">=</span><span class="n">stdout</span><span class="p">)</span>

<div class="viewcode-block" id="BwaMemAlignCommandline"><a class="viewcode-back" href="../../align.html#proto_err.align.BwaMemAlignCommandline">[docs]</a><span class="k">class</span> <span class="nc">BwaMemAlignCommandline</span><span class="p">(</span><span class="n">AbstractCommandline</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Command line wrapper for Burrows Wheeler Aligner (BWA) aln.</span>

<span class="sd">    Run a BWA alignment, equivalent to::</span>

<span class="sd">        $ bwa aln [...] &lt;in.db.fasta&gt; &lt;in.query.fq&gt; &gt; &lt;out.sai&gt;</span>

<span class="sd">    See http://bio-bwa.sourceforge.net/bwa.shtml for details.</span>

<span class="sd">    Example:</span>

<span class="sd">    &gt;&gt;&gt; from Bio.Sequencing.Applications import BwaAlignCommandline</span>
<span class="sd">    &gt;&gt;&gt; reference_genome = &quot;/path/to/reference_genome.fasta&quot;</span>
<span class="sd">    &gt;&gt;&gt; read_file = &quot;/path/to/read_1.fq&quot;</span>
<span class="sd">    &gt;&gt;&gt; output_sai_file = &quot;/path/to/read_1.sai&quot;</span>
<span class="sd">    &gt;&gt;&gt; read_group=&quot;@RG\tID:foo\tSM:bar&quot;</span>
<span class="sd">    &gt;&gt;&gt; align_cmd = BwaAlignCommandline(reference=reference_genome, read_file=read_file)</span>
<span class="sd">    &gt;&gt;&gt; print(align_cmd)</span>
<span class="sd">    bwa aln /path/to/reference_genome.fasta /path/to/read_1.fq</span>

<span class="sd">    You would typically run the command line using align_cmd(stdout=output_sai_file)</span>
<span class="sd">    or via the Python subprocess module, as described in the Biopython tutorial.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cmd</span><span class="o">=</span><span class="s">&quot;bwa&quot;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">program_name</span> <span class="o">=</span> <span class="n">cmd</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span> <span class="o">=</span> \
                <span class="p">[</span>
                    <span class="n">_StaticArgument</span><span class="p">(</span><span class="s">&quot;mem -B 3 -t 4 -O 1 -L 9&quot;</span><span class="p">),</span>
                    <span class="n">_Argument</span><span class="p">([</span><span class="s">&quot;reference&quot;</span><span class="p">],</span> <span class="s">&quot;Reference file name&quot;</span><span class="p">,</span>
                              <span class="n">filename</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">is_required</span><span class="o">=</span><span class="bp">True</span><span class="p">),</span>
                    <span class="n">_Argument</span><span class="p">([</span><span class="s">&quot;read_file&quot;</span><span class="p">],</span> <span class="s">&quot;Read file name&quot;</span><span class="p">,</span>
                              <span class="n">filename</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">is_required</span><span class="o">=</span><span class="bp">True</span><span class="p">),</span>
                    <span class="n">_Option</span><span class="p">([</span><span class="s">&quot;-k&quot;</span><span class="p">,</span> <span class="s">&quot;k&quot;</span><span class="p">],</span>
                            <span class="s">&quot;Minimum seed length. Matches shorter than INT will be missed. The alignment speed is usually insensitive to this value unless it significantly deviates 20.&quot;</span><span class="p">,</span>
                            <span class="n">checker_function</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">)),</span>
                            <span class="n">equate</span><span class="o">=</span><span class="bp">False</span><span class="p">),</span>
                    <span class="c"># _Option([&quot;-o&quot;, &quot;o&quot;],</span>
                    <span class="c">#         &quot;Maximum edit distance if the value is INT, or the fraction of missing alignments given 2% uniform base error rate if FLOAT. In the latter case, the maximum edit distance is automatically chosen for different read lengths. [0.04]&quot;,</span>
                    <span class="c">#         checker_function=lambda x: isinstance(x, (int, float)),</span>
                    <span class="c">#         equate=False),</span>
                    <span class="c"># _Option([&quot;-e&quot;, &quot;e&quot;],</span>
                    <span class="c">#         &quot;Maximum number of gap extensions, -1 for k-difference mode (disallowing long gaps) [-1]&quot;,</span>
                    <span class="c">#         checker_function=lambda x: isinstance(x, int),</span>
                    <span class="c">#         equate=False),</span>
                    <span class="c"># _Option([&quot;-d&quot;, &quot;d&quot;],</span>
                    <span class="c">#         &quot;Disallow a long deletion within INT bp towards the 3-end [16]&quot;,</span>
                    <span class="c">#         checker_function=lambda x: isinstance(x, int),</span>
                    <span class="c">#         equate=False),</span>
                    <span class="c"># _Option([&quot;-i&quot;, &quot;i&quot;],</span>
                    <span class="c">#         &quot;Disallow an indel within INT bp towards the ends [5]&quot;,</span>
                    <span class="c">#         checker_function=lambda x: isinstance(x, int),</span>
                    <span class="c">#         equate=False),</span>
                    <span class="c"># _Option([&quot;-l&quot;, &quot;l&quot;],</span>
                    <span class="c">#         &quot;&quot;&quot;Take the first INT subsequence as seed.</span>

                    <span class="c">#         If INT is larger than the query sequence, seeding will be disabled.</span>
                    <span class="c">#         For long reads, this option is typically ranged from 25 to 35 for</span>
                    <span class="c">#         -k 2. [inf]&quot;&quot;&quot;,</span>
                    <span class="c">#         checker_function=lambda x: isinstance(x, int),</span>
                    <span class="c">#         equate=False),</span>
                    <span class="c"># _Option([&quot;-k&quot;, &quot;k&quot;], &quot;Maximum edit distance in the seed [2]&quot;,</span>
                    <span class="c">#         checker_function=lambda x: isinstance(x, int),</span>
                    <span class="c">#         equate=False),</span>
                    <span class="c"># _Option([&quot;-t&quot;, &quot;t&quot;], &quot;Number of threads (multi-threading mode) [1]&quot;,</span>
                    <span class="c">#         checker_function=lambda x: isinstance(x, int),</span>
                    <span class="c">#         equate=False),</span>
                    <span class="c"># _Option([&quot;-M&quot;, &quot;M&quot;],</span>
                    <span class="c">#         &quot;Mismatch penalty. BWA will not search for suboptimal hits with a score lower than (bestScore-misMsc). [3]&quot;,</span>
                    <span class="c">#         checker_function=lambda x: isinstance(x, int),</span>
                    <span class="c">#         equate=False),</span>
                    <span class="c"># _Option([&quot;-O&quot;, &quot;O&quot;], &quot;Gap open penalty [11]&quot;,</span>
                    <span class="c">#         checker_function=lambda x: isinstance(x, int),</span>
                    <span class="c">#         equate=False),</span>
                    <span class="c"># _Option([&quot;-E&quot;, &quot;E&quot;], &quot;Gap extension penalty [4]&quot;,</span>
                    <span class="c">#         checker_function=lambda x: isinstance(x, int),</span>
                    <span class="c">#         equate=False),</span>
                    <span class="c"># _Option([&quot;-R&quot;, &quot;R&quot;],</span>
                    <span class="c">#         &quot;&quot;&quot;Proceed with suboptimal alignments if there are no more than INT equally best hits.</span>

                    <span class="c">#         This option only affects paired-end mapping. Increasing this threshold helps</span>
                    <span class="c">#         to improve the pairing accuracy at the cost of speed, especially for short</span>
                    <span class="c">#         reads (~32bp).&quot;&quot;&quot;,</span>
                    <span class="c">#         checker_function=lambda x: isinstance(x, int),</span>
                    <span class="c">#         equate=False),</span>
                    <span class="c"># _Option([&quot;-q&quot;, &quot;q&quot;],</span>
                    <span class="c">#         &quot;&quot;&quot;Parameter for read trimming [0].</span>

                    <span class="c">#         BWA trims a read down to argmax_x{\sum_{i=x+1}^l(INT-q_i)} if q_l&lt;INT</span>
                    <span class="c">#         where l is the original read length.&quot;&quot;&quot;,</span>
                    <span class="c">#         checker_function=lambda x: isinstance(x, int),</span>
                    <span class="c">#         equate=False),</span>
                    <span class="c"># _Option([&quot;-B&quot;, &quot;B&quot;],</span>
                    <span class="c">#         &quot;Length of barcode starting from the 5-end. When INT is positive, the barcode of each read will be trimmed before mapping and will be written at the BC SAM tag. For paired-end reads, the barcode from both ends are concatenated. [0]&quot;,</span>
                    <span class="c">#         checker_function=lambda x: isinstance(x, int),</span>
                    <span class="c">#         equate=False),</span>
                    <span class="c"># _Switch([&quot;-c&quot;, &quot;c&quot;],</span>
                    <span class="c">#         &quot;Reverse query but not complement it, which is required for alignment in the color space.&quot;),</span>
                    <span class="n">_Switch</span><span class="p">([</span><span class="s">&quot;-H&quot;</span><span class="p">,</span> <span class="s">&quot;H&quot;</span><span class="p">],</span>
                            <span class="s">&quot;Hardclipping off&quot;</span><span class="p">),</span>
                    <span class="c"># _Switch([&quot;-I&quot;, &quot;I&quot;],</span>
                    <span class="c">#         &quot;The input is in the Illumina 1.3+ read format (quality equals ASCII-64).&quot;),</span>
                    <span class="c"># _Switch([&quot;-b&quot;, &quot;b&quot;],</span>
                    <span class="c">#         &quot;Specify the input read sequence file is the BAM format&quot;),</span>
                    <span class="c"># _Switch([&quot;-b1&quot;, &quot;b1&quot;],</span>
                    <span class="c">#         &quot;When -b is specified, only use the first read in a read pair in mapping (skip single-end reads and the second reads).&quot;),</span>
                    <span class="c"># _Switch([&quot;-b2&quot;, &quot;b2&quot;],</span>
                    <span class="c">#         &quot;When -b is specified, only use the second read in a read pair in mapping.&quot;)</span>
                  <span class="p">]</span>
        <span class="n">AbstractCommandline</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cmd</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</pre></div></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../../index.html">proto_err 0.0.1 documentation</a> &raquo;</li>
          <li><a href="../index.html" >Module code</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Phelim Bradley.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b3.
    </div>
  </body>
</html>